### PR TITLE
Implemented new checker ``invalid-class-object``

### LINKED ISSUE #585: [easy] __class__ must be set to a class
Originally reported by: **Claudiu Popa (BitBucket: [PCManticore](http://bitbucket.org/PCManticore), GitHub: @PCManticore)**

---

```
#!python

class A: pass
A.__class__ = 1
```

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/585

### PR DESCRIPTION
**Pull request by yushao2**
## Steps

- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description

Implemented a new checker ``invalid-class-object``, which checks for cases where a non-class object is assigned to the ``__class__`` attribute

Also, refactored files in functional test suite

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Related Issue
Closes #585

### CODE REVIEW

**Comment by yushao2** [comment]
P.S.: I'm not too familiar with working with ``__class__`` myself, so the functional test suite might be incomplete

File: pylint/checkers/classes.py [hunk 1]
```diff
@@ -1379,12 +1385,22 @@ class member from outside its class (but ignore __special__
 
         self._check_protected_attribute_access(node)
 
-    def visit_assignattr(self, node):
+    def visit_assignattr(self, node: astroid.AssignAttr) -> None:
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes.py:1388]: Maybe we could add the decorator here ?

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Nice, that's an old issue you're fixing here !

**Review by Pierre-Sassoulas — APPROVED** [review]

**Review by cdce8p — COMMENTED** [review]
Why did you move the tests cases? Not sure this is really necessary. If it is, it would probably be better to do it in a separate PR.

**Comment by yushao2** [comment]
> Why did you move the tests cases? Not sure this is really necessary. If it is, it would probably be better to do it in a separate PR.

An error was being thrown that the invalid subfolder had too many test cases.

Maybe i can undo moving all test cases, and keep this in a subfolder. What do you think?

**Comment by Pierre-Sassoulas** [comment]
I think we can keep the file moved the way you did, it's not confusing. Maybe in another MR that we'll rebase this one on ?

**Comment by yushao2** [comment]
> I think we can keep the file moved the way you did, it's not confusing. Maybe in another MR that we'll rebase this one on ?

sorry, so I should leave the files as how i moved them in this MR? just to clarify

**Comment by Pierre-Sassoulas** [comment]
Yes, looks fine to me, there is a lot of functional tests starting by "invalid" so separating them again like you did make sense.

**Comment by yushao2** [comment]
> Yes, looks fine to me, there is a lot of functional tests starting by "invalid" so separating them again like you did make sense.

for now, in this MR, i separated _only those_ with > 2 files so that it would not be too heavy on the refactoring

**Review by cdce8p — APPROVED** [review]
Ok, let's merge it then!

**Pull request merged**